File: windows/src/RealtimeTranslator.Core/Settings/AppReleaseVersion.cs
using System;

namespace RealtimeTranslator.Core.Settings;

/// <summary>
/// リリースタグやアセンブリへ埋め込んだバージョン文字列を、設定画面向けの表示値へ正規化する。
/// </summary>
public static class AppReleaseVersion
{
    /// <summary>未リリース / 埋め込みなしのときの表示値。</summary>
    public const string Unpublished = "0.0.0";

    /// <summary>
    /// <c>v0.1.0</c> や <c>0.1.0+commit</c> から、設定に出すバージョンを取り出す。
    /// </summary>
    public static string DisplayValue(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return Unpublished;
        }

        var value = raw.Trim();
        var plus = value.IndexOf('+');
        if (plus >= 0)
        {
            value = value[..plus];
        }

        if (value.Length >= 2 && (value[0] is 'v' or 'V') && char.IsAsciiDigit(value[1]))
        {
            value = value[1..];
        }

        value = value.Trim();
        return string.IsNullOrEmpty(value) ? Unpublished : value;
    }
}

File: RealtimeTranslator/App/AppReleaseVersion.swift
import Foundation

/// リリースタグや Info.plist へ埋め込んだバージョン文字列を、設定画面向けの表示値へ正規化する。
enum AppReleaseVersion {
    /// 未リリース / 埋め込みなしのときの表示値。
    static let unpublished = "0.0.0"

    /// 実行中バンドルの `CFBundleShortVersionString`。
    static var current: String {
        displayValue(
            from: Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
        )
    }

    /// `v0.1.0` や `0.1.0+commit` から、設定に出すバージョンを取り出す。
    static func displayValue(from raw: String?) -> String {
        guard var value = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else {
            return unpublished
        }

        if let plus = value.firstIndex(of: "+") {
            value = String(value[..<plus])
        }

        if value.count >= 2, value.first == "v" || value.first == "V",
           let second = value.dropFirst().first, ("0"..."9").contains(second)
        {
            value.removeFirst()
        }

        value = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return value.isEmpty ? unpublished : value
    }
}
